cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/CreateSchedulesStage.java

```diff
@@ -81,54 +76,39 @@ public void process(ApplicationWithPrograms input) throws Exception {
         // hence it is useless to update the schedule
         continue;
       }
-      ProgramType programType = ProgramType.valueOfSchedulableType(newScheduleSpec.getProgram().getProgramType());
-      ProgramId programId = appId.program(programType, newScheduleSpec.getProgram().getProgramName());
 
       // if the schedule differ in schedule type then we have deleted the existing one earlier in DeleteScheduleStage.
       // create it with the new type and spec here. See CDAP-8918 for details.
       if (ScheduleType.fromSchedule(newScheduleSpec.getSchedule()) !=
         ScheduleType.fromSchedule(oldScheduleSpec.getSchedule())) {
         LOG.debug("Redeploying schedule {} with specification {} which existed earlier with specification {}",
                   entry.getKey(), newScheduleSpec, oldScheduleSpec);
-        createSchedule(programId, newScheduleSpec);
+        addProgramSchedule(appId, newScheduleSpec);
         continue;
       }
 
-      scheduler.updateSchedule(programId,
-                               newScheduleSpec.getProgram().getProgramType(),
-                               newScheduleSpec.getSchedule(), newScheduleSpec.getProperties());
-      if (oldScheduleSpec.getSchedule() instanceof TimeSchedule) {
-        programScheduler.deleteSchedule(input.getApplicationId().schedule(oldScheduleSpec.getSchedule().getName()));
-      }
-      addProgramSchedule(programId, newScheduleSpec);
+      ProgramType programType = ProgramType.valueOfSchedulableType(newScheduleSpec.getProgram().getProgramType());
+      ProgramId programId = appId.program(programType, newScheduleSpec.getProgram().getProgramName());
+
+      programScheduler.updateSchedule(Schedulers.toProgramSchedule(newScheduleSpec.getSchedule(), programId,
+                                                                   newScheduleSpec.getProperties()));
     }
 
     for (Map.Entry<String, ScheduleSpecification> entry : mapDiff.entriesOnlyOnRight().entrySet()) {
-      ScheduleSpecification scheduleSpec = entry.getValue();
-      ProgramType programType = ProgramType.valueOfSchedulableType(scheduleSpec.getProgram().getProgramType());
-      createSchedule(appId.program(programType, scheduleSpec.getProgram().getProgramName()), scheduleSpec);
+      addProgramSchedule(appId, entry.getValue());
     }
 
     // Emit the input to next stage.
     emit(input);
   }
 
-  private void createSchedule(ProgramId programId, ScheduleSpecification scheduleSpec) throws SchedulerException {
-    scheduler.schedule(programId, scheduleSpec.getProgram().getProgramType(), scheduleSpec.getSchedule(),
-                       scheduleSpec.getProperties());
-    addProgramSchedule(programId, scheduleSpec);
-  }
-
-  private void addProgramSchedule(ProgramId programId, ScheduleSpecification scheduleSpec) {
+  private void addProgramSchedule(ApplicationId appId, ScheduleSpecification scheduleSpec)
```

Comment:
Right, and I think it will be easier in`DefaultAppConfigurer#addSchedule`, we can directly add old schedules as ProgramSchedules instead of creating `ScheduleSpecification`, but this requires REST API changes to prevent test failures